services/ui-src/src/components/fields/ChoiceListField.tsx

```diff
@@ -27,22 +31,54 @@ export const ChoiceListField = ({
   const form = useFormContext();
   form.register(name);
 
+  const [fieldValues, setFieldValues] = useState<string[] | null>(
+    props.hydrate || form.getValues(name) || null
+  );
+
   const formatChoices = (choices: FieldChoice[]) =>
     choices.map((choice: FieldChoice) => {
       const choiceObject: FieldChoice = { ...choice };
       const choiceChildren = choice?.children;
       if (choiceChildren) {
-        const isNested = true;
-        const formattedChildren = formFieldFactory(choiceChildren, isNested);
+        const formattedChildren = formFieldFactory(choiceChildren, true);
         choiceObject.checkedChildren = formattedChildren;
       }
       delete choiceObject.children;
       return choiceObject;
     });
 
-  const nestedChildClasses = nested ? "nested ds-c-choice__checkedChild" : "";
+  // update local state
+  const onChangeHandler = (event: InputChangeEvent) => {
+    const clickedOption = event.target.value;
+    const isOptionChecked = event.target.checked;
+    const currentFieldValues = fieldValues || [];
+    // handle radio
+    if (type === "radio") {
+      setFieldValues([clickedOption]);
+    } else {
+      // handle checkbox
+      setFieldValues(
+        isOptionChecked
+          ? [...currentFieldValues, clickedOption]
+          : currentFieldValues.filter((value) => value !== clickedOption)
+      );
+    }
+  };
+
+  useEffect(() => {
+    // update form data
+    if (fieldValues) {
+      form.setValue(name, fieldValues, { shouldValidate: true });
+      // update DOM choices checked status
+      choices.forEach((choice: FieldChoice) => {
+        choice.checked = fieldValues.includes(choice.value);
+      });
+    }
+  }, [fieldValues]);
 
+  const nestedChildClasses = nested ? "nested ds-c-choice__checkedChild" : "";
   const parsedHint = hint && parseCustomHtml(hint);
+  const errorMessage = form?.formState?.errors?.[name]?.message;
```

Comment:
ChangeHandler, useEffect, and errorMessage relocated from CheckboxField and RadioField